docs: describe single region and multi-region deployments (#12779)
coder/coder: docs/about/architecture.md

# Architecture

The Coder deployment model is flexible and offers various components that
platform administrators can deploy and scale depending on the use cases. This
page describes possible deployments, challenges, and risks associated with them.

Learn more about our [Reference Architectures](../admin/architectures/index.md)
and platform scaling capabilities.

## Primary components

### coderd

_coderd_ is the service created by running `coder server`. It is a thin API that
connects workspaces, provisioners and users. _coderd_ stores its state in
Postgres and is the only service that communicates with Postgres.

It offers:

- Dashboard (UI)
- HTTP API
- Dev URLs (HTTP reverse proxy to workspaces)

> - Workspace Web Applications (e.g easily access `code-server`)

Reviewer: change to "access to"

##### Workload resources

- Use Terraform to deploy at least one **Coder Server Replica** per availability
  zone with _Coder Server_ instances and provisioners. High availability is
  recommended but not essential for small deployments.
- Single replica deployment is a special case that can address a
  tiny/small/proof-of-concept installation on a single virtual machine serving
  less than 100 workspace users.

**Coder workspace**

- For small deployments consider a lightweight workspace runtime like
  [Sysbox](https://github.com/nestybox/sysbox) container runtime. Learn more how
  to enable
  [docker-in-docker using Sysbox](https://asciinema.org/a/kkTmOxl8DhEZiM2fLZNFlYzbo?speed=2).

**HA Database**

- Enable replication across all availability zones.
- Monitor replication lag, node status, and resource utilization metrics.
- Implement robust backup and disaster recovery strategies to protect against
  data loss.

##### Workload supporting resources

**Load balancer**

- Distributes and load balances traffic from agents and clients to _Coder
  Server_ replicas across zones.
- Layer 7 load balancing. The load balancer can decrypt SSL traffic, and

>   re-encrypt using internal certificate.

Reviewer: change to "using an internal certificate"

> - Sessions persistence (sticky sessions) can be disabled as _Coder Server_

Reviewer: change to "Session"

  instances are stateless.

> - WebSocket and long-time connections must be supported.

Reviewer: change to "long-lived connections"

Author: Brain fart, of course long-lived :) Thanks!

**Single sign-on**

> - Integrate with existing Single Sign-On (SSO) solution used within the

Reviewer: change to "solutions"

>   organization, OAuth 2.0 or OpenID Connect standards are supported.

Reviewer: change to "organization via the supported OAuth 2.0 or OpenID Connect standards."

- Learn more about [Authentication in Coder](../admin/auth.md).

### Multi-region architecture

![Architecture Diagram](../images/architecture-multi-region.png)

#### Components

This architecture is for globally distributed developer teams using Coder
workspaces on daily basis. It features a single load balancer with regionally
deployed _Workspace Proxy Replicas_, several _Coder Server_ replicas, and _Coder
workspaces_ provisioned in different regions.

Note: The _multi-region architecture_ assumes the same deployment principles as
the _single region architecture_, but it extends them to multi region deployment

> with workspace proxies. Provisioners deployed in regions closed to developers

Reviewer: change to "Proxies are deployed in regions closest to developers to offer the fastest developer experience."

offer the fastest developer experience.

##### Workload resources

**Workspace proxy**

- Workspace proxy offers developers the option to establish a fast relay
  connection when accessing their workspace via SSH, a workspace application, or
  port forwarding.
- Dashboard connections, API calls (e.g. _list workspaces_) are not served over
  proxies.
- Proxies do not establish connections to the database.

> - Do not share authentication tokens between proxy instances.

Reviewer: Change to "Proxy instances do not share authentication tokens between one another".

Each resource may also be _persistent_ or _ephemeral_ depending on whether
they're destroyed on workspace stop.

## Deployment models

### Single region architecture

![Architecture Diagram](../images/architecture-single-region.png)

#### Components

This architecture consists of a single load balancer, several _Coder Server_

> replicas, and _Coder workspaces_ deployed in the same region.

Author: I took the inspiration about glossary from the graph, but `coderd` sounds more natural. Agree 👍

##### Workload resources

> - Use Terraform to deploy at least one **Coder Server Replica** per availability

Author: Fixed (skipped).

  zone with _Coder Server_ instances and provisioners. High availability is
  recommended but not essential for small deployments.
- Single replica deployment is a special case that can address a

>   tiny/small/proof-of-concept installation on a single virtual machine serving
>   less than 100 workspace users.

Reviewer 2: ```suggestion
  tiny/small/proof-of-concept installation on a single virtual machine. It's
  not recommended to serve more than 100 users/workspaces in this case.
```
It does make coder sound "bad" if it can't handle more than 100 workspaces with a single replica, so I though we could re-word this slightly.

Another take could be:

> If you are serving more than 100 users/workspaces, you should add more replicas.

**Coder workspace**

> - For small deployments consider a lightweight workspace runtime like

Reviewer 2: ```suggestion
- For small deployments consider a lightweight workspace runtime like the
```

  [Sysbox](https://github.com/nestybox/sysbox) container runtime. Learn more how
  to enable
  [docker-in-docker using Sysbox](https://asciinema.org/a/kkTmOxl8DhEZiM2fLZNFlYzbo?speed=2).

**HA Database**

> - Enable replication across all availability zones.

Reviewer 2: What's the purpose for this? Do we actually support DB replication? Is this something that's only available in cloud deployments/via cloud providers?

Looking outside the cloud, in PostgreSQL terms this could be a hot standby or something. But having one in each availability zone doesn't make sense since you can only use the primary with Coder. Unless we've added support for doing RO queries on a standby instance.

Reviewer 2: I think it's fine to say HA database, but I don't think we can say too much about how it's setup unless we support it in `coderd`. If we leave it as a mysterious "HA Database" without expanding on it too much, I think it's OK.

But if we do support this in some way I'm not aware, my comment may be outdated.

Like, maybe customers can achieve this in cloud easily, or maybe through manual setup of PG Bouncer, or some such, but I personally wouldn't know how to setup a HA Database for `coderd` off the top of my head. And I do know how to setup replication and hot standby's. 😄

Author: Fair point 👍 I left the mysterious _HA Database_ section, but removed references to replication.

- Monitor replication lag, node status, and resource utilization metrics.
- Implement robust backup and disaster recovery strategies to protect against
  data loss.

##### Workload supporting resources

**Load balancer**

- Distributes and load balances traffic from agents and clients to _Coder

>   Server_ replicas across zones.

Reviewer 2: ```suggestion
  Server_ replicas across availability zones.
```
Should we use consistent terminology here?

- Layer 7 load balancing. The load balancer can decrypt SSL traffic, and
  re-encrypt using an internal certificate.
- Session persistence (sticky sessions) can be disabled as _Coder Server_
  instances are stateless.
- WebSocket and long-lived connections must be supported.

**Single sign-on**

- Integrate with existing Single Sign-On (SSO) solutions used within the

>   organization organization via the supported OAuth 2.0 or OpenID Connect
>   standards.

Reviewer 2: ```suggestion
  organization via the supported OAuth 2.0 or OpenID Connect standards.
```

- Learn more about [Authentication in Coder](../admin/auth.md).

### Multi-region architecture

![Architecture Diagram](../images/architecture-multi-region.png)

#### Components

This architecture is for globally distributed developer teams using Coder
workspaces on daily basis. It features a single load balancer with regionally

> deployed _Workspace Proxy Replicas_, several _Coder Server_ replicas, and _Coder

Reviewer 2: Nit: consistent use of "_something Replicas_" and "_something_ replicas.

PS. Should we call these "Workspaces Proxies" instead of workspace proxy replicas? The distinction makes me think "workspace proxies" 1 proxy/region whereas "workspace proxy replica" makes me think N proxies/region.

workspaces_ provisioned in different regions.

Note: The _multi-region architecture_ assumes the same deployment principles as
the _single region architecture_, but it extends them to multi region deployment
with workspace proxies. Proxies are deployed in regions closest to developers to
offer the fastest developer experience.

##### Workload resources

**Workspace proxy**

- Workspace proxy offers developers the option to establish a fast relay
  connection when accessing their workspace via SSH, a workspace application, or
  port forwarding.
- Dashboard connections, API calls (e.g. _list workspaces_) are not served over
  proxies.
- Proxies do not establish connections to the database.
- Proxy instances do not share authentication tokens between one another.

##### Workload supporting resources

**Proxy load balancer**

- Distributes and load balances workspace relay traffic in a single region
  across availability zones.
- Layer 7 load balancing. The load balancer can decrypt SSL traffic, and

>   re-encrypt using internal certificate.

Reviewer 2: Similar, but missing some bullet points like websockets from the earlier section.

Author: Added items from the earlier section.
